### ISSUE TITLE
`BeEmpty()` no longer lists all items in the collection in the error message

### ISSUE THREAD

**Issue by nyctef** [fluentassertions/fluentassertions#2782]
### Description

First off: thanks for the awesome library! We've been using it for a few years now and it's been a huge boon to our tests :)

This is kind of a https://xkcd.com/1172/ situation, but I thought it'd be at least worth writing down to see what other people think.

Previously, using `foo.Should().BeEmpty()` on a non-empty collection would produce a result like "expected foo to be empty, but was {1, 2, 3}". Now it produces an error like "expected foo to be empty, but found at least one item {1}". 

This is annoying in the case where the test needs to be re-run multiple times to discover more unexpected items in the collection, especially if the test is slow to run or can only be run on the CI server. An example in our case is checking for dlls that are part of the installed product which aren't authenticode-signed.

### Reproduction Steps

```csharp
[Test]
public async Task Check_things_are_valid()
{
    var things = new List<string> { "fine", "PROBLEM", "other_fine", "OTHER_PROBLEM" };

    var problems = new List<string>();
    foreach (var thing in things)
    {
        if (!await SomeLongWindedProcessToCheckIfThingIsValid(thing))
        {
            problems.Add(thing);
        }
    }
    problems.Should().BeEmpty();
}

private Task<bool> SomeLongWindedProcessToCheckIfThingIsValid(string thing) =>
    Task.FromResult(thing == thing.ToLowerInvariant());
```

### Expected behavior

The error message from the test should list all of the problems so they can be fixed in one go.

As a potential compromise in the case of infinite collections, it could maybe list up to some max number of items?

### Actual behavior

The error message from the test only lists one problem at a time, meaning we have to fix one at a time, push and wait for a CI build to see if there's another problem or not

### Regression?

This appears to have been an intentional change in https://github.com/fluentassertions/fluentassertions/pull/2530

### Known Workarounds

`problems.Should().Equal([]);` prints out all of the problems (but looks wrong: someone will come along and say "why don't you just use BeEmpty?")

### Configuration

using FluentAssertions 6.12.1 and .NET 8.0.403

### Are you willing to help with a pull-request?

Maybe? I don't really know the code that well but can try if we agree on a change

**Comment by dennisdoomen** [comment]
I understand your problem, but as you've discovered yourself, we intentionally decided to make `BeEmpty` not go any further than that. Honoring your request would mean we break the behavior again. For now, we decide to keep it as is.

**Comment by dennisdoomen** [comment]
Yeah, I could live with that one.

**Comment by dennisdoomen** [comment]
@nyctef Would you be willing to try to implement this?

**Comment by nyctef** [comment]
Sure, I can give it a go - I should have some spare time on Friday so I'll try picking it up then.

Just to be clear on my end, I think these are the requirements:
- can only enumerate once for both checking the condition and printing the assertion message
- can only enumerate a finite number of items to avoid breaking on infinite collections

the thing I'm not sure about is so long as we satisfy both of those, does it matter if the number of items materialized is like 5 or 500?

**Comment by jnyrup** [comment]
I agree on the requirements.

On the second parts my gut feeling is to pick at most 10.
How many items were there in the collection in your test suite?

**Comment by nyctef** [comment]
looks like `EnumerableValueFormatter` has a default limit of 32 items, so maybe we could just lean on that? I assume it would make the limit user-configurable as well

I think the trick at that point would be to save the first element of the collection (if it exists) and then create a new enumerable with that element at the front for formatting the message

**Comment by jnyrup** [comment]
It it was me I would just raise the current `Take(1)` to e.g. `Take(10)`, adjust the tests, add release notes and call it a day.

... If my test started failing with more than 10 items, I hope those would give a strong enough indication of where to look for the bug.

**Comment by nyctef** [comment]
Yeah, and the majority of the time I think that would be reasonable - I fully admit our use case is a bit unusual, where each item in the collection is a problem that needs to be addressed separately 😅

**Comment by dennisdoomen** [comment]
> It it was me I would just raise the current `Take(1)` to e.g. `Take(10)`, adjust the tests, add release notes and call it a day.
> 
> ... If my test started failing with more than 10 items, I hope those would give a strong enough indication of where to look for the bug.

Let's do that then.
